### ISSUE TITLE
Suggestions In Improving Code

### ISSUE THREAD

**Issue by CaveNightingale** [Andy-K-Sparklight/ACH#32]
I have skimmed through all your C++ source files and found some problems.

----

https://github.com/Andy-K-Sparklight/Ari/blob/aafc16c36ca8b594d952302d890841a39df8b4d6/src/Main.cc#L116
Here, `params` is not cJSON_Delete()ed.

----

https://github.com/Andy-K-Sparklight/Ari/blob/aafc16c36ca8b594d952302d890841a39df8b4d6/src/util/Commons.cc#L80
Here, the return value should be `ssize_t`, since `size_t` indicates unsigned.

----

https://github.com/Andy-K-Sparklight/Ari/blob/aafc16c36ca8b594d952302d890841a39df8b4d6/src/sys/Schedule.cc#L100
https://github.com/Andy-K-Sparklight/Ari/blob/aafc16c36ca8b594d952302d890841a39df8b4d6/src/sys/Schedule.cc#L147
Is this really thread safe? How do you make sure worker thread don't delete `tx`?

----

https://github.com/Andy-K-Sparklight/Ari/blob/aafc16c36ca8b594d952302d890841a39df8b4d6/src/sys/Schedule.cc#L134
Resource Leak. C++ is not Java, you need to `detach` a thread if you don't want to `join`, or it will not release its resource and wait for other thread to `join`.

**Comment by Andy-K-Sparklight** [comment]
I'll explain them in detail.

The undeleted json object and wrong return value: fixed

For the implementation of `runOnWorkerThread` , the executor `x` in `WorkerData` has no access to pointer `tx`. Also, even if they can, we won't do that.

The last one: `uv_create_thread` provided `uv_thread_join` but NO correponding version of `pthread_detach`. Also we cannot treat `uv_thread_t` as `pthread_t` because:

> ...but this is an implementation detail, avoid depending on it to always be true.

For this I'll open another issue to fix.

**Comment by Andy-K-Sparklight** [comment]
Besides, it would be better if using *the code* rather than *your code*. I commited most part of the codebase but there are other contributors.

**Comment by CaveNightingale** [comment]
> I'll explain them in detail.
> 
> The undeleted json object and wrong return value: fixed
> 
> For the implementation of `runOnWorkerThread` , the executor `x` in `WorkerData` has no access to pointer `tx`. Also, even if they can, we won't do that.
> 
> The last one: `uv_create_thread` provided `uv_thread_join` but NO correponding version of `pthread_detach`. Also we cannot treat `uv_thread_t` as `pthread_t` because:
> 
> > ...but this is an implementation detail, avoid depending on it to always be true.
> 
> For this I'll open another issue to fix.

I don't agree with you, if you add some `thrd_yield()` before joining to simulate giving up the rest CPU times.
like:
```
void
runOnWorkerThread(std::function<void()> func, bool join)
{
  auto tx = new WorkerData;
  tx->x = func;
  uv_thread_create(
      &tx->tid,
      [](void *a) -> void {
        auto t = (WorkerData *)a;
        t->x();
        delete t;
      },
      tx);
  thrd_yield();
  thrd_yield();
  thrd_yield();
  thrd_yield();
  if(join)
    {
      uv_thread_join(&tx->tid);
    }
}
```
and then call in `main`:
```
Alicorn::Sys::runOnWorkerThread([]() -> void {
    LOG("aaa");
  }, true);
```
You will probably get a `SIGSEGV`.

**Comment by CaveNightingale** [comment]
Thread safety is a serious thing.

**Comment by Andy-K-Sparklight** [comment]
I'm using the following test code:
```c++
void
runOnWorkerThread(std::function<void()> func, bool join)
{
  std::thread *tx = new std::thread([func, tx]() -> void {
    func();
    delete tx;
  });
  for(int i = 0; i < 1000; i++)
    {
      std::this_thread::yield();
    }
  if(join)
    {
      tx->join();
    }
  else
    {
      tx->detach();
    }
}
```
and

```c++
Sys::runOnWorkerThread([]() -> void { LOG("Everything was right!"); }, true);
```

I ran it for 20 times under both GDB and MINGW shell, no error was reported.
Please also run the test code on a Linux OS to confirm.

**Comment by Andy-K-Sparklight** [comment]
I've just used DrMemory to scan the memory and I couldn't find anything related to memory leaking or access violation.

**Comment by CaveNightingale** [comment]
> I'm using the following test code:
> 
> ```c++
> void
> runOnWorkerThread(std::function<void()> func, bool join)
> {
>   std::thread *tx = new std::thread([func, tx]() -> void {
>     func();
>     delete tx;
>   });
>   for(int i = 0; i < 1000; i++)
>     {
>       std::this_thread::yield();
>     }
>   if(join)
>     {
>       tx->join();
>     }
>   else
>     {
>       tx->detach();
>     }
> }
> ```
> 
> and 
> 
> ```c++
> Sys::runOnWorkerThread([]() -> void { LOG("Everything was right!"); }, true);
> ```
> 
> I ran it for 20 times under both GDB and MINGW shell, no error was reported. Please also run the test code on a Linux OS to confirm.

Terribly bad solution.
Let's see what you are deleting via GDB. (You are deleting 0x0)
![截图 2023-01-29 22-40-32](https://user-images.githubusercontent.com/51874645/215334084-0a513c64-819e-4748-9c64-7144faa0579d.png)


And there is a warning from compiler.
```
../src/sys/Schedule.cc: In function ‘void Alicorn::Sys::runOnWorkerThread(std::function<void()>, bool)’:
../src/sys/Schedule.cc:110:1: warning: ‘tx’ may be used uninitialized [-Wmaybe-uninitialized]
  110 | }
      | ^
../src/sys/Schedule.cc:91:16: note: ‘tx’ was declared here
   91 |   std::thread *tx = new std::thread([func, tx]() -> void {
      |                ^~
```

Are you kidding me?

**Comment by Andy-K-Sparklight** [comment]
Please issue a pull request if there is a solution to fix.
I've received the same warning too, by declare the pointer first and then assign it will solve it.

**Comment by CaveNightingale** [comment]
> 

There is an easy solution.
```
void
runOnWorkerThread(std::function<void()> func, bool join)
{
  std::thread tx(func);
  if(join)
    {
      tx.join();
    }
  else
    {
      tx.detach();
    }
}
```
